l20n/migrate/transforms.py

```diff
@@ -0,0 +1,247 @@
+# coding=utf8
+
+import l20n.format.ast as FTL
+
+
+def evaluate(ctx, node):
+    def eval_node(subnode):
+        if isinstance(subnode, Transform):
+            return subnode(ctx)
+        else:
+            return subnode
+
+    return node.map(eval_node)
+
+
+class Transform(FTL.Node):
+    def __call__(self, ctx):
+        raise NotImplementedError
+
+
+class SOURCE(Transform):
+    """Declare the source translation to be migrated with other transforms."""
+    def __init__(self, path, key):
+        self.path = path
+        self.key = key
+
+    def __call__(self, ctx):
+        return ctx.get_source(self.path, self.key)
+
+
+class COPY(Transform):
+    """Copy the translation value from the given source.
+
+    The translation must be a simple value without interpolations nor plural
+    variants.  All \uXXXX will be converted to the literal characters they
+    encode.
+
+    HTML entities are left unchanged for now.  They're use might or might not
+    be intentional.  Consider the following example in which `&amp;` could be
+    replaced with the literal `&`:
+
+        Privacy &amp; History
+
+    vs. these two examples where the HTML encoding should be preserved:
+
+        Erreur&nbsp;!
+        Use /help &lt;command&gt; for more information.
+
+    """
+
+    # XXX Perhaps there's a strict subset of HTML entities which must or must
+    # not be replaced?
+
+    def __init__(self, source):
+        self.source = source
+
+    def __call__(self, ctx):
+        return FTL.Pattern(
+            source=None,
+            elements=[FTL.TextElement(self.source)],
+            quoted=(
+                self.source.startswith(' ') or self.source.endswith(' ')
+            )
+        )
+
+
+class EXTERNAL(Transform):
+    """Create an FTL placeable with the external argument `name`
+
+    This is a common use-case when joining translations with CONCAT.
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    def __call__(self, ctx):
+        external = FTL.ExternalArgument(self.name)
+        elements = [FTL.Placeable([external])]
+
+        return FTL.Pattern(None, elements, quoted=False)
+
+
+class REPLACE(Transform):
+    """Replace various placeables in the translation with FTL placeables.
+
+    The original placeables are defined as keys on the `replacements` dict.
+    For each key the value is defined as a list of FTL Expressions to be
+    interpolated.
```

Comment:
Also, funky placeholders like [this one](https://hg.mozilla.org/releases/l10n/mozilla-aurora/it/file/default/browser/chrome/browser/browser.properties#l243)
```
Impossibile connettersi a %2$S in questo momento.%1$0.S
```